### PR TITLE
Is fully commutative for Coxeter group elements

### PR DESCRIPTION
**Pull request by fchapoton**
### :books: Description

This is adding a method "is_fully_commutative" to all Coxeter group elements.

On the way, introduce iterators over the sets of reduced words of a given element.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fchapoton** [comment]
Maybe one could try instead to add a method "`is_fully_commutative`" to the `BraidOrbit` classes

File: src/sage/categories/coxeter_groups.py [hunk 1]
```diff
@@ -251,6 +251,53 @@ def braid_group_as_finitely_presented_group(self):
...
+            r"""
+            Iterate over the braid orbit of a word ``word`` of indices.
+
+            The input word does not need to be a reduced expression of
+            an element.
+
+            INPUT:
+
+            - ``word``: a list (or iterable) of indices in
+              ``self.index_set()``
```
Comment (tscrim) [inline src/sage/categories/coxeter_groups.py]:
```suggestion
            - ``word`` -- a list (or iterable) of indices in
              ``self.index_set()``
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Can we add a test showing things works for a Coxeter group with a "be_careful" indexing set?

Also, personally, I think we should expunge the `# optional` markers until they are actually used (in particular, can be reasonably tested against).

**Comment by fchapoton** [comment]
> Can we add a test showing things works for a Coxeter group with a "be_careful" indexing set?

Hello Travis, thanks for the idea. Would you please suggest a simple way to create such a group ?

(I prefer not to mess with these damn markers)

**Comment by tscrim** [comment]
There seems to be a bit of bug in the `CoxeterMatrixGroup` input ignoring the index set when passed something that can be interpreted as a Coxeter type.

Well, we can at least work around that by working with a Coxeter group for a very generic matrix:
```python
sage: W = CoxeterGroup(CM, index_set=('a', Permutation([2,1]), -2), implementation="reflection")
sage: W.index_set()
('a', [2, 1], -2)
```

**Comment by fchapoton** [comment]
Thanks. Is my new doctest enough ?

**Comment by fchapoton** [comment]
should be better now

**Review by tscrim — APPROVED** [review]
Thank you. This is good.

**Pull request merged**